Author:LINCC Frameworks team

Last updated: July 08, 2025

# Possible AGN problem 2 - crossmatch large number of AGN

In [1]:
# Get MiliQuas catalog from https://quasars.org/milliquas.htm
# transform into something that lsdb can ingest, like pandas df
# load into lsdb with lsdb.from_dataframe function
# crossmatch with DP1 object collections 
# get all AGN that you can find
# crossmatch with DP1 dia_object collections 
# get all AGN that you can find
# What is the difference in the number of objects and dia_objects you have found?

In [2]:
# !wget https://quasars.org/milliquas.fits.zip

In [3]:
# astropy.Table.read can handle .gz compression but not .zip
# !unzip milliquas.fits.zip

In [4]:
from astropy.table import Table

mdata = Table.read("milliquas.fits")
pmdata = mdata.to_pandas()
pmdata

RA        DEC                          NAME     TYPE  \
0          0.000629  35.517844  b'SDSS J000000.15+353104.2 '  b'Q   '   
1          0.000657 -20.074344  b'UVQSJ000000.15-200427.7  '  b'AX  '   
2          0.001450  31.057037  b'SDSS J000000.33+310325.3 '  b'Q   '   
3          0.001527  27.732283  b'SDSS J000000.36+274356.2 '  b'Q   '   
4          0.001535   7.064129  b'SDSS J000000.36+070350.8 '  b'Q   '   
...             ...        ...                           ...      ...   
1021795  359.999121  28.954727  b'SDSS J235959.78+285717.0 '  b'Q   '   
1021796  359.999293  34.720858  b'SDSS J235959.83+344315.0 '  b'Q   '   
1021797  359.999633   3.268621  b'SDSS J235959.90+031606.9 '  b'Q   '   
1021798  359.999746  20.721087  b'SDSS J235959.94+204315.8 '  b'Q   '   
1021799  359.999954  34.747133  b'SDSS J235959.98+344449.6 '  b'Q   '   

              RMAG       BMAG COMMENT     R     B      Z       CITE  \
0        17.930000  19.000000  b'pG '  b'-'  b'-'  0.845  b'DR16Q '   
1        17.900000  18.459999  b'pG '  b'-'  b'1'  0.291  b'UVQS  '   
2        19.950001  21.290001  b'jN '  b'-'  b'-'  2.035  b'DR16Q '   
3        21.750000  21.910000  b'g  '  b'-'  b'-'  1.771  b'DR16Q '   
4        21.870001  22.110001  b'g  '  b'-'  b'-'  1.574  b'DR16Q '   
...            ...        ...     ...   ...   ...    ...        ...   
1021795  21.360001  21.350000  b'g+N'  b'-'  b'-'  2.452  b'DR16Q '   
1021796  19.780001  20.420000  b'gG '  b'-'  b'-'  3.109  b'DR16Q '   
1021797  18.610001  18.889999  b'gG '  b'-'  b'1'  1.233  b'DR16Q '   
1021798  20.270000  20.350000  b'gG '  b'-'  b'-'  2.010  b'DR16Q '   
1021799  20.570000  20.879999  b'g  '  b'-'  b'-'  2.375  b'DR16Q '   

             ZCITE                      XNAME                      RNAME  \
0        b'DR16  '  b'                      '  b'                      '   
1        b'UVQS  '  b'LSXPS J000000.1-200426'  b'                      '   
2        b'DR16  '  b'                      '  b'                      '   
3        b'DR16  '  b'                      '  b'                      '   
4        b'DR16  '  b'                      '  b'                      '   
...            ...                        ...                        ...   
1021795  b'DR16Q '  b'                      '  b'                      '   
1021796  b'DR16Q '  b'                      '  b'                      '   
1021797  b'DR16  '  b'                      '  b'                      '   
1021798  b'DR16  '  b'                      '  b'                      '   
1021799  b'DR16Q '  b'                      '  b'                      '   

                             LOBE1                      LOBE2  
0        b'                      '  b'                      '  
1        b'                      '  b'                      '  
2        b'                      '  b'                      '  
3        b'                      '  b'                      '  
4        b'                      '  b'                      '  
...                            ...                        ...  
1021795  b'                      '  b'                      '  
1021796  b'                      '  b'                      '  
1021797  b'                      '  b'                      '  
1021798  b'                      '  b'                      '  
1021799  b'                      '  b'                      '  

[1021800 rows x 16 columns]

In [5]:
import lsdb

m_cat = lsdb.from_dataframe(pmdata, ra_column="RA", dec_column="DEC")
m_cat

/astro/users/dtj1s/.conda/envs/dtj1s-py3.12/lib/python3.12/site-packages/lsdb/loaders/dataframe/dataframe_catalog_loader.py:108: RuntimeWarning: from_dataframe is not intended for large datasets. Consider using hats-import: https://hats-import.readthedocs.io/
  warnings.warn(


Dask NestedFrame Structure:
                                  RA              DEC             NAME             TYPE            RMAG            BMAG          COMMENT                R                B               Z             CITE            ZCITE            XNAME            RNAME            LOBE1            LOBE2
npartitions=12                                                                                                                                                                                                                                                                                  
0                    double[pyarrow]  double[pyarrow]  binary[pyarrow]  binary[pyarrow]  float[pyarrow]  float[pyarrow]  binary[pyarrow]  binary[pyarrow]  binary[pyarrow]  float[pyarrow]  binary[pyarrow]  binary[pyarrow]  binary[pyarrow]  binary[pyarrow]  binary[pyarrow]  binary[pyarrow]
288230376151711744               ...              ...              ...              ...             ...             ...              ...              ...              ...             ...              ...              ...              ...              ...              ...              ...
...                              ...              ...              ...              ...             ...             ...              ...              ...              ...             ...              ...              ...              ...              ...              ...              ...
3170534137668829184              ...              ...              ...              ...             ...             ...              ...              ...              ...             ...              ...              ...              ...              ...              ...              ...
3458764513820540928              ...              ...              ...              ...             ...             ...              ...              ...              ...             ...              ...              ...              ...              ...              ...              ...
Dask Name: nestedframe, 3 expressions
Expr=MapPartitions(NestedFrame)

In [6]:
dp1_obj = lsdb.open_catalog("/epyc/data3/hats/catalogs/dp1/object_collection")
dp1_obj

Dask NestedFrame Structure:
                           coord_dec    coord_decErr         coord_ra     coord_raErr       g_psfFlux    g_psfFluxErr        g_psfMag     g_psfMagErr       i_psfFlux    i_psfFluxErr        i_psfMag     i_psfMagErr        objectId           patch       r_psfFlux    r_psfFluxErr        r_psfMag     r_psfMagErr          refBand         refFwhm     shape_flag        shape_xx        shape_xy        shape_yy           tract       u_psfFlux    u_psfFluxErr        u_psfMag     u_psfMagErr                x            xErr                y       y_psfFlux    y_psfFluxErr        y_psfMag     y_psfMagErr            yErr       z_psfFlux    z_psfFluxErr        z_psfMag     z_psfMagErr                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                objectForcedSource
npartitions=389                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
9147936743096320     double[pyarrow]  float[pyarrow]  double[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  int64[pyarrow]  int64[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  string[pyarrow]  float[pyarrow]  bool[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  int64[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  double[pyarrow]  float[pyarrow]  double[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  float[pyarrow]  nested<coord_ra: [double], coord_dec: [double], visit: [int64], detector: [int16], band: [string], psfFlux: [float], psfFluxErr: [float], psfFlux_flag: [bool], psfDiffFlux: [float], psfDiffFluxErr: [float], psfDiffFlux_flag: [bool], pixelFlags_bad: [bool], pixelFlags_cr: [bool], pixelFlags_crCenter: [bool], pixelFlags_edge: [bool], pixelFlags_interpolated: [bool], pixelFlags_interpolatedCenter: [bool], pixelFlags_nodata: [bool], pixelFlags_saturated: [bool], pixelFlags_saturatedCenter: [bool], pixelFlags_suspect: [bool], pixelFlags_suspectCenter: [bool], invalidPsfFlag: [bool]

In [7]:
from dask.distributed import Client
# Because I'm on arnor and I know I can
client = Client(n_workers=4, memory_limit="32GB")

/astro/users/dtj1s/.conda/envs/dtj1s-py3.12/lib/python3.12/site-packages/distributed/node.py:187: UserWarning: Port 8787 is already in use.
Perhaps you already have a cluster running?
Hosting the HTTP server on port 45481 instead
  warnings.warn(


In [8]:
m_x_dp1_obj = m_cat.crossmatch(dp1_obj, radius_arcsec=0.1)
m_x_dp1_obj

Dask NestedFrame Structure:
                    RA_from_lsdb_dataframe DEC_from_lsdb_dataframe NAME_from_lsdb_dataframe TYPE_from_lsdb_dataframe RMAG_from_lsdb_dataframe BMAG_from_lsdb_dataframe COMMENT_from_lsdb_dataframe R_from_lsdb_dataframe B_from_lsdb_dataframe Z_from_lsdb_dataframe CITE_from_lsdb_dataframe ZCITE_from_lsdb_dataframe XNAME_from_lsdb_dataframe RNAME_from_lsdb_dataframe LOBE1_from_lsdb_dataframe LOBE2_from_lsdb_dataframe coord_dec_object_lc coord_decErr_object_lc coord_ra_object_lc coord_raErr_object_lc g_psfFlux_object_lc g_psfFluxErr_object_lc g_psfMag_object_lc g_psfMagErr_object_lc i_psfFlux_object_lc i_psfFluxErr_object_lc i_psfMag_object_lc i_psfMagErr_object_lc objectId_object_lc patch_object_lc r_psfFlux_object_lc r_psfFluxErr_object_lc r_psfMag_object_lc r_psfMagErr_object_lc refBand_object_lc refFwhm_object_lc shape_flag_object_lc shape_xx_object_lc shape_xy_object_lc shape_yy_object_lc tract_object_lc u_psfFlux_object_lc u_psfFluxErr_object_lc u_psfMag_object_lc u_psfMagErr_object_lc      x_object_lc  xErr_object_lc      y_object_lc y_psfFlux_object_lc y_psfFluxErr_object_lc y_psfMag_object_lc y_psfMagErr_object_lc  yErr_object_lc z_psfFlux_object_lc z_psfFluxErr_object_lc z_psfMag_object_lc z_psfMagErr_object_lc                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      objectForcedSource_object_lc     _dist_arcsec
npartitions=142                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
9147936743096320           double[pyarrow]         double[pyarrow]          binary[pyarrow]          binary[pyarrow]    

In [9]:
%%time
m_x_dp1_obj_c = m_x_dp1_obj.compute()
m_x_dp1_obj_c

/astro/users/dtj1s/.conda/envs/dtj1s-py3.12/lib/python3.12/site-packages/distributed/client.py:3383: UserWarning: Sending large graph of size 59.60 MiB.
This may cause some slowdown.
Consider loading the data with Dask directly
 or using futures or delayed objects to embed the data into the graph without repetition.
See also https://docs.dask.org/en/stable/best-practices.html#load-data-with-dask for more information.
  warnings.warn(


CPU times: user 2.53 s, sys: 141 ms, total: 2.67 s
Wall time: 19.4 s


RA_from_lsdb_dataframe  DEC_from_lsdb_dataframe  \
_healpix_29                                                            
9583286184997455                  37.784846                 6.401708   
9589207111743847                  38.493228                  6.42792   
...                                     ...                      ...   
2530266678998616679               53.066504               -27.577352   
2531113080423514090               40.491485               -34.346719   

                         NAME_from_lsdb_dataframe TYPE_from_lsdb_dataframe  \
_healpix_29                                                                  
9583286184997455     b'LAMOSTJ023108.36+062406.2'                  b'Q   '   
9589207111743847     b'LAMOSTJ023358.37+062540.4'                  b'QR  '   
...                                           ...                      ...   
2530266678998616679  b'WCDFS J033215.96-273438.4'                  b'A   '   
2531113080423514090  b'Q J0241-3420             '                  b'Q   '   

                     RMAG_from_lsdb_dataframe  BMAG_from_lsdb_dataframe  \
_healpix_29                                                               
9583286184997455                    19.389999                     19.73   
9589207111743847                        18.83                     20.26   
...                                       ...                       ...   
2530266678998616679                 22.940001                 24.559999   
2531113080423514090                 18.290001                 18.959999   

                    COMMENT_from_lsdb_dataframe R_from_lsdb_dataframe  \
_healpix_29                                                             
9583286184997455                         b'pG '                  b'-'   
9589207111743847                         b'jG '                  b'1'   
...                                         ...                   ...   
2530266678998616679                      b'd  '                  b'n'   
2531113080423514090                      b'jG '                  b'-'   

                    B_from_lsdb_dataframe  Z_from_lsdb_dataframe  ...  \
_healpix_29                                                       ...   
9583286184997455                     b'-'                  1.404  ...   
9589207111743847                     b'-'                  3.932  ...   
...                                   ...                    ...  ...   
2530266678998616679                  b'n'                  0.742  ...   
2531113080423514090                  b'-'                   1.43  ...   

                    y_psfFluxErr_object_lc y_psfMag_object_lc  \
_healpix_29                                                     
9583286184997455                      <NA>               <NA>   
9589207111743847                      <NA>               <NA>   
...                                    ...                ...   
2530266678998616679             936.135132          21.656466   
2531113080423514090                   <NA>               <NA>   

                    y_psfMagErr_object_lc yErr_object_lc z_psfFlux_object_lc  \
_healpix_29                                                                    
9583286184997455                     <NA>       0.006627        56795.382812   
9589207111743847                     <NA>       0.008974        82239.742188   
...                                   ...            ...                 ...   
2530266678998616679              0.129329        0.06311         5985.897949   
2531113080423514090                  <NA>       0.003418                <NA>   

                    z_psfFluxErr_object_lc  z_psfMag_object_lc  \
_healpix_29                                                      
9583286184997455                319.930939           19.514217   
9589207111743847                318.370789           19.112295   
...                                    ...                 ...   
2530266678998616679             110.244308           21.957176   
253111

# Possible AGN problem 3 - run a function across AGN lightcurves

## Computing `drw_fit` in memory

At this point we have `m_x_dp1_obj` in memory, and it's easy to examine and process it.

In [10]:
from nested_pandas.utils import count_nested

m_x_dp1_obj_ct = count_nested(m_x_dp1_obj_c, "objectForcedSource_object_lc")
long_light = m_x_dp1_obj_ct.query("n_objectForcedSource_object_lc > 200")
long_light

RA_from_lsdb_dataframe  DEC_from_lsdb_dataframe  \
_healpix_29                                                            
1450134596509057692               95.133821               -25.254856   
2415759207242145357                58.88646               -48.859599   
...                                     ...                      ...   
2530253631113069183               52.966718               -27.722697   
2530253698219385731               52.990831               -27.702488   

                         NAME_from_lsdb_dataframe TYPE_from_lsdb_dataframe  \
_healpix_29                                                                  
1450134596509057692  b'PKS 0618-252             '                  b'QR2 '   
2415759207242145357  b'3HSPJ035532.7-485134     '                  b'B2  '   
...                                           ...                      ...   
2530253631113069183  b'WCDFS J033152.03-274321.7'                  b'Q   '   
2530253698219385731  b'WCDFS J033157.79-274208.8'                  b'AX  '   

                     RMAG_from_lsdb_dataframe  BMAG_from_lsdb_dataframe  \
_healpix_29                                                               
1450134596509057692                     18.58                 17.860001   
2415759207242145357                      18.6                     20.82   
...                                       ...                       ...   
2530253631113069183                 22.950001                     23.43   
2530253698219385731                     22.25                 23.530001   

                    COMMENT_from_lsdb_dataframe R_from_lsdb_dataframe  \
_healpix_29                                                             
1450134596509057692                      b'pG '                  b'1'   
2415759207242145357                      b'jG '                  b'1'   
...                                         ...                   ...   
2530253631113069183                      b'd  '                  b'n'   
2530253698219385731                      b'd  '                  b'-'   

                    B_from_lsdb_dataframe  Z_from_lsdb_dataframe  ...  \
_healpix_29                                                       ...   
1450134596509057692                  b'-'                    1.9  ...   
2415759207242145357                  b'1'                    0.5  ...   
...                                   ...                    ...  ...   
2530253631113069183                  b'n'                    1.5  ...   
2530253698219385731                  b'-'                  0.668  ...   

                    y_psfMag_object_lc y_psfMagErr_object_lc yErr_object_lc  \
_healpix_29                                                                   
1450134596509057692          17.994703              0.002773       0.001976   
2415759207242145357          18.958101              0.003162       0.009979   
...                                ...                   ...            ...   
2530253631113069183          22.514696              0.150656       0.093441   
2530253698219385731          21.071037              0.035332       0.030747   

                    z_psfFlux_object_lc z_psfFluxErr_object_lc  \
_healpix_29                                                      
1450134596509057692          310360.875             120.127457   
2415759207242145357        93192.304688              88.527161   
...                                 ...                    ...   
2530253631113069183         2360.625732              67.332741   
2530253698219385731         9889.001953              70.140358   

                    z_psfMag_object_lc  z_psfMagErr_object_lc  \
_healpix_29                                                     
1450134596509057692          17.670332                0.00042   
2415759207242145357          18.976549               0.001031   
...                                ...                    ...   
2530253631113069183          22.967432               0.030977   
253025369821

In [11]:
from eztao.ts import drw_fit

In [12]:
%%time
def do_fit(lc_nest_in):
    lc_nest = lc_nest_in.dropna()
    coeffs = drw_fit(
        lc_nest["midpointMjdTai"].to_numpy(),
        lc_nest["psfMag"].to_numpy(),
        lc_nest["psfMagErr"].to_numpy())
    drw_0, drw_1 = coeffs
    return { "drw_0": drw_0, "drw_1": drw_1 }

fits = long_light.reduce(do_fit, "objectForcedSource_object_lc")
fits

CPU times: user 2min 29s, sys: 263 ms, total: 2min 30s
Wall time: 15.2 s


drw_0     drw_1
_healpix_29                            
1450134596509057692  0.499093  0.002283
2415759207242145357  0.947583  0.005800
...                       ...       ...
2530253631113069183  0.224432  0.001120
2530253698219385731  1.131532  0.002094

[478 rows x 2 columns]

## Computing `drw_fit` on the catalog

So far so good!  But suppose that we had a lot more than 478 rows and we wanted to leverage the cluster.

Now we want to transplant our in-memory computations confidently to the catalog.  The big problem here is "confidently".

In [15]:
# Step 1: count the number of elements in each nest and only keep those with > 200 points.

from nested_pandas.utils import count_nested

def count_points(pts):
    # Asked to count `objectForcedSource_object_lc`, this will add a column called `n_objectForcedSource_object_lc`
    return count_nested(pts, "objectForcedSource_object_lc")

counted_cat = m_x_dp1_obj.map_partitions(count_points)
long_light_cat = counted_cat.query("n_objectForcedSource_object_lc > 200")

### Transplant direct to `lsdb.catalog.Catalog.reduce`?

In [19]:
# Use the previous function directly.
# Hideous `meta=` required.
fit_cat = long_light_cat.reduce(do_fit, "objectForcedSource_object_lc", meta={"drw_0": float, "drw_1":float})
fit_cat

Dask NestedFrame Structure:
                       drw_0    drw_1
npartitions=142                      
9147936743096320     float64  float64
9570149208162304         ...      ...
...                      ...      ...
2531132941744996352      ...      ...
2531137339791507456      ...      ...
Dask Name: nestedframe, 8 expressions
Expr=MapPartitions(NestedFrame)

#### This actually worked!

In [20]:
fit_cat.compute()

/astro/users/dtj1s/.conda/envs/dtj1s-py3.12/lib/python3.12/site-packages/distributed/client.py:3383: UserWarning: Sending large graph of size 59.74 MiB.
This may cause some slowdown.
Consider loading the data with Dask directly
 or using futures or delayed objects to embed the data into the graph without repetition.
See also https://docs.dask.org/en/stable/best-practices.html#load-data-with-dask for more information.
  warnings.warn(


drw_0     drw_1
_healpix_29                            
1450134596509057692  0.499093  0.002283
2415759207242145357  0.947583  0.005800
...                       ...       ...
2530253631113069183  0.224432  0.001120
2530253698219385731  1.131532  0.002094

[478 rows x 2 columns]

### But what about using `.map_partitions`?

One big difference is that the argument given to the user-provided function is the base row.

In [34]:
import pandas as pd
def do_fit_with_map_1(nf_in):
    nf = nf_in.copy()
    # nf["objectForcedSource_object_lc"] = nf["objectForcedSource_object_lc"].dropna()
    drw_0, drw_1 = 0.0, 0.0
    if True:  # len(nf) > 0:
        try:
            t = nf["objectForcedSource_object_lc.midpointMjdTai"].to_numpy()
            coeffs = drw_fit(
                t,
                nf["objectForcedSource_object_lc.psfMag"].to_numpy(),
                nf["objectForcedSource_object_lc.psfMagErr"].to_numpy())
            drw_0, drw_1 = coeffs
        except Exception:
            pass
    return pd.DataFrame(
        [
            {
                "drw_0": drw_0,
                "drw_1": drw_1
            }
        ]
    )

In [35]:
fit_mp_cat = long_light_cat.map_partitions(do_fit_with_map_1)
fit_mp_cat

Dask NestedFrame Structure:
                       drw_0    drw_1
npartitions=142                      
9147936743096320     float64  float64
9570149208162304         ...      ...
...                      ...      ...
2531132941744996352      ...      ...
2531137339791507456      ...      ...
Dask Name: nestedframe, 8 expressions
Expr=MapPartitions(NestedFrame)

In [36]:
fit_mp_cat.compute()

/astro/users/dtj1s/.conda/envs/dtj1s-py3.12/lib/python3.12/site-packages/distributed/client.py:3383: UserWarning: Sending large graph of size 59.74 MiB.
This may cause some slowdown.
Consider loading the data with Dask directly
 or using futures or delayed objects to embed the data into the graph without repetition.
See also https://docs.dask.org/en/stable/best-practices.html#load-data-with-dask for more information.
  warnings.warn(


drw_0  drw_1
0     0.0    0.0
0     0.0    0.0
..    ...    ...
0     0.0    0.0
0     0.0    0.0

[142 rows x 2 columns]

Ridiculous.  No data, and wrong number of rows.  It's per-partition.  What?